### PR TITLE
[FIX] web: return updated list of fields in export template

### PR DESCRIPTION
**Pull request by prec-odoo**
When user access template at the time of export with deleted field. The traceback will be generated.

To reproduce the issue(any model, here- `account.move.line`):

- Install `account_accountant` module
- Go to Settings > Technical > Database Structure > Fields
- Create a new field with model as `Journal Item` and save
- Go to Accounting > Miscellaneous > Journal Items
- Select any record in list view and click on export and generate a new export template with newly created field
- Go to `ir.model.fields` and delete that field
- Go to `Journal Items` and select that template while export

Traceback in sentry -
```
KeyError: 'tax_audit'
  File "odoo/http.py", line 2123, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1699, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1726, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1927, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 234, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 190, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 716, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/export.py", line 373, in namelist
    return [
  File "addons/web/controllers/export.py", line 374, in <listcomp>
    {'name': field['name'], 'label': fields_data[field['name']]}
```

When a field gets deleted from `ir.model.fields` but it does not get deleted from export template. And selecting that template to export the records will lead to traceback.

https://github.com/odoo/odoo/blob/59669e9943158e51dcbb9ae69ad758df8f7c7976/addons/web/controllers/main.py#L1815-L1818

sentry-4331986723

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by prec-odoo** [comment]
Hello @odoo/sentry_reviewers  ,

Can you please look into this PR?

Thank you!

File: addons/web/controllers/main.py [hunk 1]
```diff
@@ -1814,7 +1814,7 @@ def namelist(self, model, export_id):
 
         return [
             {'name': field['name'], 'label': fields_data[field['name']]}
-            for field in export_fields_list
+            for field in export_fields_list if field['name'] in self.fields_get(model)
```
Comment (yelizariev) [inline addons/web/controllers/main.py:1817]: You should put `self.fields_get(model)` in a variable to avoid the call every time

**Review by yelizariev — COMMENTED** [review]
It looks as an edge case, but on the other hand the patch doesn't introduce much noise. So, make the requested changes and then we decide

**Comment by prec-odoo** [comment]
> It looks as an edge case, but on the other hand the patch doesn't introduce much noise. So, make the requested changes and then we decide

@yelizariev changes has been done.

**Review by Julien00859 — COMMENTED** [review]
The fix makes it work but doesn't solve the underlying issue, those `ir.exports.line` records should be removed on cascade with the de-facto related `ir.model.fields` or the removal of the `ir.model.fields` should be prevented as long as it is still used in an export template.

Looking at the definition of the `ir.exports.line` model, I'm surprised that there is a complete lack of referential integrity, i.e. that the fields are `fields.Char()` instead of `fields.Many2one('ir.model.fields')` with an extra related for the name. IMO we should fix the model but that's not an option for stable...

On one hand as the field is removed, it is only logic to remove it from the template (using this kind of runtime verification in stable). On the other hand those templates might not have been created by the same person who removed the field and I'm not too sure that *silently* discarding the field would always be ok. We could raise an `UserError` to let the user know that this template is broken but the UI requires that you select a template in order to remove it thus it would be impossible to remove it... Sending a warning via the bus is not an option either as web doesn't depend on bus...

Maybe something in the like of...

```py
all_fields = request.env['ir.exports.line'].browse(export['export_fields'])
export_fields = all_fields.filtered(self.fields_get(model).__contains__)
ko_fields = all_fields - export_fields
if ko_fields:
    # ir.exports.line lack a ondelete=cascade foreign key on ir.model.fields
    logger.warning(...)
    ko_fields.unlink()
```

Maybe I'm overthinking it...

Thank you for the complete steps to reproduce the bug, can you add a test? In the `test_xlsx_export` module

cc @bouvyd

**Comment by prec-odoo** [comment]
> The fix makes it work but doesn't solve the underlying issue, those `ir.exports.line` records should be removed on cascade with the de-facto related `ir.model.fields` or the removal of the `ir.model.fields` should be prevented as long as it is still used in an export template.
> 
> Looking at the definition of the `ir.exports.line` model, I'm surprised that there is a complete lack of referential integrity, i.e. that the fields are `fields.Char()` instead of `fields.Many2one('ir.model.fields')` with an extra related for the name. IMO we should fix the model but that's not an option for stable...
> 
> On one hand as the field is removed, it is only logic to remove it from the template (using this kind of runtime verification in stable). On the other hand those templates might not have been created by the same person who removed the field and I'm not too sure that _silently_ discarding the field would always be ok. We could raise an `UserError` to let the user know that this template is broken but the UI requires that you select a template in order to remove it thus it would be impossible to remove it... Sending a warning via the bus is not an option either as web doesn't depend on bus...
> 
> Maybe something in the like of...
> 
> ```python
> all_fields = request.env['ir.exports.line'].browse(export['export_fields'])
> export_fields = all_fields.filtered(self.fields_get(model).__contains__)
> ko_fields = all_fields - export_fields
> if ko_fields:
>     # ir.exports.line lack a ondelete=cascade foreign key on ir.model.fields
>     logger.warning(...)
>     ko_fields.unlink()
> ```
> 
> Maybe I'm overthinking it...
> 
> Thank you for the complete steps to reproduce the bug, can you add a test? In the `test_xlsx_export` module
> 
> cc @bouvyd

@Julien00859 changes has been done.

**Comment by bouvyd** [comment]
> Maybe I'm overthinking it...

@Julien00859 
A bit :)

I don't think we should raise, if a field has been deleted, it has been deleted. Whilst this use case uses manual deletion, I suspect the same could occur with an uninstallation, etc. This can happen easily, no need to make a fuss about it.

I agree that having integrity would be nice, but given hor rare a case it is i'm fine with this on the long term as well. What you suggest is more correct than what is suggested here, but the suggestion is still correct as well. Just more patchy. I can live with patchy in an area of the framework that is seldom used.


@prec-odoo small nitpicking, can you make the log warning more explicit?
> Field %s not found for saved ir.exports of model %s, deleting

seems more explicit (explains what is wrong and what was done)

File: addons/web/controllers/main.py [hunk 2]
```diff
@@ -1807,7 +1807,14 @@ def get_fields(self, model, prefix='', parent_name= '',
     def namelist(self, model, export_id):
         # TODO: namelist really has no reason to be in Python (although itertools.groupby helps)
         export = request.env['ir.exports'].browse([export_id]).read()[0]
-        export_fields_list = request.env['ir.exports.line'].browse(export['export_fields']).read()
+        exported_fields = request.env['ir.exports.line'].browse(export['export_fields'])
+
+        for invalid_fields in exported_fields.filtered(lambda f: f.name not in self.fields_get(model)):
```
Comment (Julien00859) [inline addons/web/controllers/main.py:1812]: can you move the `self.fields_get(model)` outside of the lambda, in a variable, we don't want to call fields_get() on every field

**Review by Julien00859 — CHANGES_REQUESTED** [review]

File: addons/web/controllers/main.py (hunk 2 above)
Comment (prec-odoo) [reply addons/web/controllers/main.py]: > can you move the `self.fields_get(model)` outside of the lambda, in a variable, we don't want to call fields_get() on every field

@Julien00859 , changes has been done.

**Review by Julien00859 — APPROVED** [review]
Thank you :)

@robodoo r+

**Comment by Julien00859** [comment]
@prec-odoo looks like we have a failing test due to the new warning, can you take a look? Maybe we should add a `self.assertLogs`or a `odoo.tools.mute_logger` inside of the failing test.

**Comment by prec-odoo** [comment]
> @prec-odoo looks like we have a failing test due to the new warning, can you take a look? Maybe we should add a `self.assertLogs`or a `odoo.tools.mute_logger` inside of the failing test.

@Julien00859 changes has been done.

**Review by Julien00859 — APPROVED** [review]
@robodoo r+

**Comment by chienandalu** [comment]
With this change subfields in export profiles are being dropped. It's quite a breaking change for lots of users. It should be rolled back IMO

**Pull request closed without merging**